OC-3474: Add reindex command for Chef Server

christophermaier opened the pull request:
Reindexing is now a server-side function, and not exposed through the REST API.

Commit:
Add support for server-side reindexing

Commit:
Couple of Dialyzer fixes

seth commented on src/chef_reindex.erl:
```diff
@@ -0,0 +1,246 @@
+%% Unfortunately, current code for "bulk retrieval" of the
+%% 'serialized_object' column from relevant tables in the database
+%% does not return this database ID.  Until Sqerl is able to perform
+%% ad-hoc SQL queries (a feature that should arrive soon!), it is
```
As it happens, adhoc queries have landed on sqerl master. I don't think we should change our approach at this time, however.

christophermaier replied:
Agreed.  This would be a nice, self-contained section of Erchef to experiment with them, though!

seth replied:
Indeed. I still think we should wrap this up for Chef 11. But it would be a good place to try out some alternate approaches which could then be factored into how bulk_get is implemented. Since the adhoc stuff is still new, I'd rather not have Chef 11 depend on it at this time.

christophermaier replied:
Yeah, that's what I meant... I should have been a little more clear.  I blame sleepiness.

seth commented on src/chef_reindex.erl:
```diff
@@ -0,0 +1,246 @@
+%% TODO: This is hard-coded into the chef_wm base_state
+%% record right now; this needs to go into the app's configuration.
+%% This function should ultimately retrieve the value from there.
+get_batch_size() -> 5.
```
This is available as `application:get_env(chef_wm, bulk_fetch_batch_size)`. We actually duplicate the config and put it in chef_db as well. But I think it would be better to pull from chef_wm since that's where this code is and at least slightly cleaner than hard-coding.

seth commented on src/chef_reindex.erl:
```diff
@@ -0,0 +1,246 @@
+    chef_db:data_bag_item_name_id_dict(Ctx, OrgId, DataBagName);
+create_name_id_dict(Ctx, OrgId, Index) when is_atom(Index) ->
+    %% Find the correct 2-arg function in the chef_db module
+    Fun = list_to_existing_atom(atom_to_list(Index) ++ "_name_id_dict"),
```
I think a helper function that matches on `Index` and does the call might be more clear? Or perhaps move that into chef_db as we've done with create and update? So the chef_db signature would be `chef_db:create_name_id_dict(Ctx, Index, OrgId)`

The multiple fun heads result in more typing, but I think you get better type transparency.

seth commented on src/chef_reindex.erl:
```diff
@@ -0,0 +1,246 @@
+%% batch of database IDs.
+%%
+%% Just returns a list of 'ok' atoms... could just as easily return a
+%% single 'ok'.
```
What do you think about having this return `ok` and then matching that in `batch_reindex/6` so that we crash early if rabbit is down or we encounter some other error along the way?

seth commented:
Great commentary throughout. The shape of it is quite clear. Just the few minor things commented above.

Commit:
Get batch size from app config

Commit:
More granular pattern matching for indexing

Should allow us to bail earlier if e.g., something's wrong with the indexing queue.

Conflicts:
	src/chef_reindex.erl

christophermaier commented:
I've made the suggested changes.  A lot was changed in chef_db to support this, so be sure to check there, too.  Much more clean this way, I feel.

seth commented:
chef_db looks good. Please MWR.
